### ISSUE TITLE
[BUG] [Gcloud] Can't launch cheap instances because of unlimited quota 

### ISSUE THREAD

**Issue by Raimund58** [state-of-the-art/BlendNet#139]
**Issue description:**
I think Gcloud changed their way to display "preemptible instances". 
I have unlimited quota on a new created project but the addon is thinking it can not run cheap instances.
Google says somehow `Unlimited==0`... ¯\\_(ツ)_/¯
(Or you detect it wrong?)
This disables any chance to launch chap instances.

![2021-02-06 12_30_59-Window](https://user-images.githubusercontent.com/20732516/107118957-a0f03280-6884-11eb-94a1-6e1edf19ad9f.png)

https://github.com/state-of-the-art/BlendNet/blob/725085ebb8c3bd3b9691418df181402441c2c0e1/BlendNet/providers/gcp/__init__.py#L232

**Comment by rabits** [comment]
Hi @Raimund58 , as far I know preemptible CPUs are disabled for the new projects and you have to go and increase quota to get them. I don't see the limit on your screenshot - just the usage, and I'm kinda belive if blendnet saying that preemptible cpus quota is zero :)

Could you please execute the next command in terminal (you can replace `us-central1` to your region):
```
$ gcloud compute regions describe us-central1 | grep -A2 -B2 PREEMPTIBLE_CPUS
```
You will see something like:
```
  usage: 0.0
- limit: 1000.0
  metric: PREEMPTIBLE_CPUS
  usage: 0.0
- limit: 1.0
```
If not - please remove the cmd part `| grep -A2 -B2 PREEMPTIBLE_CPUS`, try again and try to find PREEMPTIBLE_CPUS there.

So this way you will find the actual quota limit, or you can go and try to get the web version here: https://cloud.google.com/compute/quotas

**Comment by Raimund58** [comment]
Hi @rabits ,
Please look again at the screenshot, the Limit is `Unlimited`  :)

Note that I am not a new gcloud user. I already sunk 600€ into it. But I didn't touch it in almost two years. The Quota site design looked different back then.

I did execute `gcloud compute regions describe us-central1`
Short output:
```
- limit: 20.0
  metric: REGIONAL_INSTANCE_GROUP_MANAGERS
  usage: 0.0
- limit: 0.0
  metric: PREEMPTIBLE_CPUS
  usage: 0.0
- limit: 1.0
  metric: NVIDIA_K80_GPUS
  usage: 0.0
  ```
Long output:
[full_command_output.txt](https://github.com/state-of-the-art/BlendNet/files/5937733/full_command_output.txt)

**Comment by rabits** [comment]
I think it's kinda bug in the UI, as far you can see: the limit for you is 0 - means 0 vCPU cores available. So please request some - and you will get some :)

**Comment by Raimund58** [comment]
@rabits Well, then why can I book cheap instances then?

**Comment by Raimund58** [comment]
![2021-02-06 23_20_34-Google Cloud Platform](https://user-images.githubusercontent.com/20732516/107130878-2817c780-68d2-11eb-9ae9-e5cd83d234d1.png)

**Comment by Raimund58** [comment]
Btw, after I removed this code bit it works flawless 
https://github.com/state-of-the-art/BlendNet/blob/725085ebb8c3bd3b9691418df181402441c2c0e1/BlendNet/providers/gcp/__init__.py#L232
(I still get warnings but at least it lets me tick the `cheap instances` box.

**Comment by rabits** [comment]
Oh, this is the new info - so your project was created long time ago when by default the users get the unlimited number of preemptible cpus :) But, you know, your case is rare... Not sure how I can detect this unlimited thing out of API, because can't reproduce the state you have... If there is a way - that will be great, and I can implement it in source code.

For example - if you can locate where those "Unlimited" values are coming from, for example via firefox/chrome dev console - there should be some json response from google api - that will be helpful.

**Comment by Raimund58** [comment]
Sorry, I'm not dev enough to find the values you are looking for :/
But I can offer you a remote connection to my machine so you can find it your self ;)

Btw, the account is two years old, but the project "BlendNet" was created yesterday.

**Comment by rabits** [comment]
Yeah, maybe those limits are tightened to the account. But for sure I can't accept your generous offer - it's too dangerous.

I checked my own quotas - and have the same "Unlimited" for my quota) And this is weird, because previously there was some limit as I remember it... So I will check how to get the actual values here.

**Comment by rabits** [comment]
Actually, on your original screenshot there is no `us-central1` - could you please find it on the page? I found on my page and it's the same as the output from API:
![screenshot](https://user-images.githubusercontent.com/1093323/107131744-14e30880-688e-11eb-9c8e-0e719161ea6a.png)

**Comment by Raimund58** [comment]
Not sure how a connection over Teamviewer can be "too dangerous", but anyway:

![2021-02-07 00_19_42-Quota metric details – IAM   Admin – BlendNet – Google Cloud Platform](https://user-images.githubusercontent.com/20732516/107131837-62856280-68da-11eb-871f-b9d2485afadf.png)
![2021-02-07 00_19_52-Quota metric details – IAM   Admin – BlendNet – Google Cloud Platform](https://user-images.githubusercontent.com/20732516/107131839-63b68f80-68da-11eb-832f-e94ba5c0aadf.png)


What Quotas do you get if you create a new test project?

**Comment by rabits** [comment]
Nono, I mean exactly the region `us-central1`, not the zones with suffixes `-a` to `-f` - could you please find it?

**Comment by Raimund58** [comment]
![2021-02-07 00_35_11-Window](https://user-images.githubusercontent.com/20732516/107132077-61553500-68dc-11eb-9e03-3a88222d78e7.png)

**Comment by rabits** [comment]
Ok, so I have to ask to make sure we did not miss anything - is this instance you showed before are in `us-central1` region?

**Comment by Raimund58** [comment]
Yes, I did set it as default region

**Comment by Raimund58** [comment]
What Quotas do you get if you create a new test project?

**Comment by rabits** [comment]
You see - now it looks like a bug on Google side (good bug, but it will be closed some day). If your quota limit for the region is 0 - that means you have no CPUs there. It's going downwards:
* Zone quota PREEMPTIBLE_CPUS
* Region quota PREEMPTIBLE_CPUS
* Zone quota CPUS
* Region quota CPUS
* Overall cpu quota CPUS

So even if you have unlimited cpus on the zone side - it still will be limeted by the upper level quotas. So I think they just made a mistake due to their software update.

**Comment by Raimund58** [comment]
Not sure, if I can follow you there.
In my understanding only the limit of the
Zone quota CPUS
Region quota CPUS
Overall cpu quota CPUS
matters.
I had never issues with a "preemptible problem".

Rabit <notifications@github.com> schrieb am So., 7. Feb. 2021, 01:07:

**Comment by rabits** [comment]
If I got your idea correctly - you think that preemptible CPUs are not limited by PREEMPTIBLE_CPUS quota?

**Comment by Raimund58** [comment]
That is my theory, yes.
(Sorry for my bad wording overall, I'm not native)

**Comment by rabits** [comment]
So you see - there is 2 issues:
* We have a special quota that is named `PREEMPTIBLE_CPUS` - why do you think it is not (supposed to be) used if it's exist?
* Without this quota I was not able to run the preemptible CPUs before, so it's in conflict with my previous experience.

So in order to resolve them - I will try to run preemptible instance on the project, where such quota is set to 0 (like in your case) and if it will work - will contact google support to figure out the rules - maybe the rules was changed and now PREEMPTIBLE_CPUS is legacy...

**Comment by rabits** [comment]
Can confirm - with N-instances I can create preemptible even with 0 in PREEMPTIBLE_CPUS on the region... Checked with 64 cores - it was limited by CPUS quota.

**Comment by rabits** [comment]
With N2 and E2 instances - the same behavior. So I'm going to contact google to clarify this weird behavior.

**Comment by rabits** [comment]
Created a ticket for gcloud folks: https://issuetracker.google.com/issues/179608117

**Comment by rabits** [comment]
So the first answer they gave me is quite misleading, but it looks like the logic under those quotas is more complicated than I thought... But need to get confirmation before preparing some changes in BlendNet logic.

**Comment by rabits** [comment]
The previous ticket was closed an noone checking the updates, so I created another one to confirm: https://issuetracker.google.com/issues/180255557

**Comment by rabits** [comment]
So the answer was on the different page here: https://cloud.google.com/compute/quotas#quotas_for_preemptible_resources :
```
However, if your project does not have preemptible quota, you can still use the regular quota to launch preemptible resources.
```
So the fix should not be hard)

**Comment by rabits** [comment]
Ok, just released v0.3.8 with the fix. Now it's not disabling cheap agents and calculates them properly.

**Comment by Raimund58** [comment]
Thank you for investigating 💯
